Update the docs for Allflex Product Mapper (#9)
lic-nz/allflex-product-mapper: README.md

## Usage

To call this mapper, pass in an `ordered product`, for an Allflex product and optionally a formatter. 

There are three kinds of formatter dependencies you can optionally inject:
  * The `NilFormatter` does nothing with the ordered product and simply returns it back. It is used if no formatter parameter is provided.
  * The `XmlFormatter` returns Allflex XML.  Not the complete order, mind, rather the XML for an individual product.  For marked tag that will be a `<MarkedOrderLine>`, for an accessory, that will be a `<BlankOrderLine>`.  Check out `./spec/shared/examples/allflex_xml.rb` for examples. You will see the various XML components being generated here using the data computed eg `OrderLine`
  * `XmlTypeSplittingFormatter`: Returns the Allflex XML and also indicates whether it is a `MarkedOrderLine` or `BlankOrderLine` product to help the correct element around each OrderLine from the external service that consumes the mapper.

Example of calling the mapper with the default NilFormatter

```ruby
result = Allflex::ProductMapper::Mapper.new.(ordered_product)
```

Example of calling the mapper with a formatter option

```ruby

> result = ::Allflex::ProductMapper::Mapper.new.(ordered_product,::Allflex::ProductMapper::XmlTypeSplittingFormatter)

Reviewer: ```suggestion
result = ::Allflex::ProductMapper::Mapper.new.(ordered_product, ::Allflex::ProductMapper::XmlTypeSplittingFormatter)
```

product.value_or  # Allflex::ProductMapper::Product
```

There are three kinds of formatter dependencies you can optionally inject:

>   * The `NilFormatter` does what it says.  

Reviewer: Not very friendly. Maybe explain why you would pass this in?

  * The `XmlFormatter` returns Allflex XML.  Not the complete order, mind, rather the XML for an individual product.  For marked tag that will be a `<MarkedOrderLine>`, for an accessory, that will be a `<BlankOrderLine>`.  Check out `./spec/shared/examples/allflex_xml.rb` for examples. You will see the various XML components being generated here using the data computed eg `OrderLine`

>   * `XmlTypeSplittingFormatter`: Returns the Allflex XML and also indicates whether it is a `MarkedOrderLine` or `BlankOrderLine` product to help the correct element around each OrderLine from the external service that consumes the mapper.

Reviewer: > to help the correct element around each OrderLine

Not sure what this means. Can you reword to make it clearer?

```ruby

> product = Allflex::ProductMapper::Mapper.new.(ordered_product ::Allflex::ProductMapper::XmlTypeSplittingFormatter)

Reviewer: ```suggestion
result = ::Allflex::ProductMapper::Mapper.new.(ordered_product, ::Allflex::ProductMapper::XmlTypeSplittingFormatter)
```

```

## Configuring Allflex Products

The Mapper is NOT configuration-free.  There is no algorithmic method to map from an LIC definition of an Allflex product, and the Allflex product defined in the Allflex systems.  Mores the pity!

Therefore, the mapping occurs in the following places:

> + `lib/allflex/product_mapper/manufacturing_definition.rb`: Defines the mappings that will translate the OrderedProduct into an Allflex-mappable form.  The manufacturing definition is mostly applied by `lib/allflex/product_mapper/product_builder.rb`

Reviewer: Make these relative links e.g. [manufacturing_definition.rb](lib/allflex/product_mapper/manufacturing_definition.rb)

> + `lib/allflex/product_mapper/allflex_product.rb`: This is essentially the Allflex Product value object.  The `product_builder` injects all the mappings into this object, which is then used by any formatter provided.  It also has the responsibility for generating the Allflex product code, which is derived from the product configuration.  So, in this class, you'll also find configuration necessary to construct the product code.  The defined codes are from Allflex API documentation at: `https://allflex.co.nz:8443/api/`

Reviewer: Maybe break this up into sections? Mapping, Tag Example, Accessor Example, Mapping Process In Detail.

> A HDX Premium Calf Pack is a single marked product in LIC, with 2 size variants (medium and large). It maps to a single manufacturing definition. This is always the case; the FulfilmentProduct to ManufacturingDefinition is 1..1. For this kind of definition the id field is insignificant--because we use the `AllflexProduct` id builder in this case.

Reviewer: ```suggestion
A HDX Premium Calf Pack is a single marked product in LIC that comes in two size variants (medium and large).

The pack always maps to a single manufacturing definition. A `FulfilmentProduct` to `ManufacturingDefinition` is `1..1`. For this kind of definition the id field is insignificant. We use the `AllflexProduct` id builder in this case.
```

  marked_tag_id_mappings: {
    "allflex-premium-calf-tag-female" => {
      "urn:lic:product:size:medium" => MEDIUM_TAMPERPROOF,
      "urn:lic:product:size:large" => LARGE_TAMPERPROOF,
      "config_hacks" => {"urn:lic:configreq:birth_id_range" => [HAS_LOGO_HACK]}
    },
    "allflex-premium-calf-tag-male" => {
      "urn:lic:product:size:medium"=>MEDIUM_MALE,
      "urn:lic:product:size:large"=>LARGE_MALE
    },
    "allflex-hdx-birth-id-eid-tag-female" => {
      "id"=> H_PERF_HDX_TAMPERPROOF,
      "config_hacks" => {"urn:lic:configreq:birth_id_range" => [HAS_LOGO_HACK]}
    },
    "allflex-hdx-birth-id-eid-tag-male" => LONG_STEM_MALE_BUTTON,
  }
}
```

However this is not the case when mapping accessories as the id field is mandatory to get the `ItemCode` needed by the schema for accessories.

```ruby
{
  id: "00194",
  product_slug: "allflex-tag-pen"
}
```

The general process works like this:

> 1) An instance of `Mapper` class is created

Reviewer: These can also be relative links to classes e.g. [Mapper](lib/allflex/product_mapper/mapper.rb)

2) A new instance of the passed in formatter is created e.g. `XmlFormatter` and passes in data that is manipulated using the input ordered product.

> 3) `OrderedProduct` wraps the LIC product JSON form.  This object provides all methods necessary to extract components from the LIC product form

Reviewer: Use either `OrderedProduct` or ordered product consistently.

Also add this clarification (maybe with a link to a fixture example?:
the LIC Ordered Product JSON format known as LIC product form

> 3) Using this ordered product, the `Fabricator` is called. Its responsibility is to inject the manufacturing definition into the pipeline.  It uses the ordered product's slug to find a product definition with a matching product_slug in `ManufacturingDefinition`. If a matched is found this product definition gets returned.

Reviewer: ```suggestion
3) Using this ordered product, the `Fabricator` is called. It's responsibility is to inject the manufacturing definition into the pipeline.  It uses the ordered product's slug to find a product definition with a matching product_slug in `ManufacturingDefinition`. If a match is found this product definition gets returned.
```

> 4) The `Fabricator` calls `ProductBuilder` is then called with the ordered product and matched manufacturing definition. The bulk of the work happens in this class. First it builds an instance of `AllflexProduct` using the input data. Methods from the `OrderedProduct` class is called to help achieve this.

Reviewer: ```suggestion
4) The `Fabricator` then calls `ProductBuilder` with the ordered product and matched manufacturing definition. The bulk of the work happens in this class. First it builds an instance of `AllflexProduct` using the input data. Methods from the `OrderedProduct` class are called to achieve this.
```

6) The `MarkingsBuilder` analyses the ordered product's configuration (in particular the config_requirement's URNs) to call the right builder -> `MgmtNumberBuilder`, `CustomTextBuilder`, `BirthIdSeqBuilder`, `BirthIdPrefixBuilder`, `BirthIdBuilder`, `BlankBuilder`, `EidRemovalBuilder` and `PrintTypeBuilder`. Any config hacks are also found here too.
5) The `AllflexProduct` builds the components contained inside the product using the product slugs in the `marked_tag_mappings` contained in the matching product definition -> the components then call the various builders -> `MarkingsBuilder`, `ColourBuilder`(it matches the URN of the colour) and fetches the size from the input ordered product data.
7) Each of these builders have an associated marking class they use to process the configuration data i.e. (`MgmtNumberMarking`, `CustomTextMarking`, `BirthIdSeqMarking`, `BirthIdPrefixMarking`, `BlankMarking`, `EidDoNotRemoveLogo`, `PrintTypeMarking`). These markings classes are all an instance of a Dry::Struct class and declare attributes for the marking as well as the name_token which appears in `AllflexProduct` definitions.

> As this is not a straightforward process, sometimes the data in the input ordered product JSON is not sufficient to meet the Allflex schema requirements so we have had to implement a number of hacks into this mapper. Namely the following:

Reviewer: Make this a seperate section maybe "Mapper hacks"
